tests/components/ccm15/test_config_flow.py

```diff
@@ -0,0 +1,132 @@
+        "host": "1.1.1.1",
+        "port": 80,
+    }
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_host(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock
+) -> None:
+    """Test we get the form."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with patch(
+        "ccm15.CCM15Device.CCM15Device.async_test_connection",
+        return_value=False,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
+    assert len(mock_setup_entry.mock_calls) == 0
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "ccm15.CCM15Device.CCM15Device.async_test_connection",
+        side_effect=CannotConnect,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+            },
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_form_unexpected_error(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "ccm15.CCM15Device.CCM15Device.async_test_connection",
+        side_effect=42,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+            },
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "unknown"}
+
+
+async def test_duplicate_host(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.config_entries._async_abort_entries_match",
+        side_effect=data_entry_flow.AbortFlow("already_configured"),
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+            },
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "unknown"}
```

Comment:
You can test this by creating a MockConfigEntry with the same hostname and adding it to hass. I recently added this for Suez Water.

Reply:
I could not get it to work with MockEntry, the test fails becuase a socket connection is attempted. Besides, this this test is good enugh as we dont really wnat to test homeassistant.config_entries._async_abort_entries_match, that is a dependency that is fair to mock.

```
async def test_duplicate_host(hass: HomeAssistant, mock_setup_entry: AsyncMock) -> None:
    """Test we handle duplicated entries."""

    entry = MockConfigEntry(
        domain=DOMAIN,
        unique_id="1.1.1.1",
        data=MOCK_DATA,
    )
    entry.add_to_hass(hass)

    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"], MOCK_DATA
    )

    assert result2["type"] == FlowResultType.FORM
    assert result2["reason"] == "already_configured"
```

Reply:
We do want to test with the real async abort entries match since the data can maybe change in the future.

Lemme take a look

Reply:
Wait that what you just posted doesn't work?

Reply:
> Wait that what you just posted doesn't work?

What I posted results in the test failling because a socket connection was made.

Reply:
bump

Reply:
Ignore the bump, GitHub didn't show your response yet.

I'll have a look tomorrow.